README.md
# esKju's jQuery ScrollFlow Plugin

## What is it?

Enhance your website by fancy effects based on the user's scroll index. Pop-in, pop-out, sliding, sticky notes and much more. The plugin supports mobile devices and eldered browsers.
It was built using the jQuery library. Licensed under MIT and GPL licenses.

## Features

+ Eased animations for incremental mouse scroll-wheels
+ Adds fancy effects to your homepage
+ Customizable trough settings and CSS
+ Highly compatible
+ Highly customizable
+ Uses CSS3 transitions by default

## How to use

### 1. doctype

Make sure you are using valid DOCTYPE. This is required for ScrollFlow to look and function correctly.

```
<!DOCTYPE html>
```

### 2. include files

Loading jQuery from CDN (Content Delivery Network) is recommended. 
Include all ScrollFlow JavaScript files in addition.

```
<script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script src="js/eskju.jquery.scrollflow.js"></script>
```

### 3. html markup

Add "scrollflow" and at least one of the following CSS classes to the desired object(s): slide-top, slide-left, slide-right, slide-bottom, -pop, -opacity

```
<h1 class="scrollflow -slide-top -opacity">Slide-Top with opacity fading</h1>
<p class="scrollflow -pop -opacity">Pop-Effect</p>
```

### 4. fire plugin

If you are not familiar with jQuery, please, read this tutorial for beginners.

```
$( document ).ready( function( )
{
	new ScrollFlow(); 
} );
```


## Options

Easy plugin, easy life. Configure things just to taste :)

```
$( document ).ready( function( )
{
	ScrollFlow( {
		your: "option",
		will: "be",
		the: "best"
	});
} );
```

| Option | Type | Default | Description |
|---------------------|---------|-------|---------------------------------------------------|
| durationOnLoad | int | 0 | Easing duration onLoad (important for page refresh) |
| durationOnResize | int | 250 | Easing duration on window resize |
| durationOnScroll | int | 500 | Easing duration on scroll |


## HTML Attributes

Customize single HTML objects

| Option | Type | Default | Description |
|---------------------|---------|-------|---------------------------------------------------|
| data-scrollflow-start | int | 25 | Define the scroll amount (in % of screen height) once the object is visible to start the animation |
| data-scrollflow-distance | int| 50 | Define the scroll amount (in % of screen height) once the object is visible to end the animation |


## Author & Help

For more information visit the author's page:

+ <http://www.cwdesigns.de> esKju's Playground
+ <http://www.cwdesigns.de/eskju-jquery-scrollflow.html> esKju's ScrollFlow Project Page


CHANGELOG.md
# Changelog

## Version 1.0.0

+ Basic plugin

js/eskju.jquery.scrollflow.js
/*
 * jQuery ScrollFlow plugin
 *
 * Copyright (c) 2015 Christian Witte
 * licensed under MIT license.
 *
 * https://github.com/eskju/eskju-jquery-scrollflow
 *
 * Version: 1.0.0
 * 
 * Licensed under MIT license.
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software 
 * and associated documentation files (the "Software"), to deal in the Software without restriction, 
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, 
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
 * subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in all copies or substantial 
 * portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT 
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

	$( document ).ready( function()
	{
		new ScrollFlow();
	});

	$.fn.ScrollFlow = function( options )
	{
		new ScrollFlow( options );
	}
	
	ScrollFlow = function( options )
	{
		this.init( options );
	}
	
	$.extend( ScrollFlow.prototype,
	{
		init : function( options )
		{
			this.options = $.extend(
			{
				useMobileTimeouts: true, // not supported yet
				mobileTimeout: 100, // not supported yet
				durationOnLoad: 0,
				durationOnResize: 250,
				durationOnScroll: 500
			}, options );
			
			this.lastScrollTop = 0;
			this.bindScroll();
			this.bindResize();
			this.update( this.options.durationOnLoad );
		},
		
		bindScroll : function()
		{
			var $this = this;
			
			$( window ).scroll( function()
			{
			 	$this.update();
			});

			$( window ).bind( "gesturechange", function()
			{
			 	$this.update();
			});
		},
		
		bindResize: function()
		{
			var $this = this;
			
			$( window ).resize( function()
			{
			 	$this.update( $this.options.durationOnResize );
			});
		},
		
		update: function( forcedDuration )
		{
			var $this = this;
			
			winHeight = $( window ).height();
			scrollTop = $( window ).scrollTop();
			
			$( ".scrollflow" ).each( function( key, obj )
			{
				objOffset = $( obj ).offset();
				objOffsetTop = parseInt( objOffset.top );
				
				// request object settings
				effectDuration = $this.options.durationOnScroll;
				effectDuration = typeof( forcedDuration ) != "undefined" ? forcedDuration : effectDuration;
				effectiveFromPercentage = ( !isNaN( parseInt( $( obj ).attr( "data-scrollflow-start" ) / 100 ) ) ? parseInt( $( obj ).attr( "data-scrollflow-start" ) ) / 100 : -0.25 );
				scrollDistancePercentage = ( !isNaN( parseInt( $( obj ).attr( "data-scrollflow-distance" ) / 100 ) ) ? parseInt( $( obj ).attr( "data-scrollflow-distance" ) ) / 100 : 0.5 );
				effectiveFrom = objOffsetTop - winHeight * ( 1 - effectiveFromPercentage );
				effectiveTo = objOffsetTop - winHeight * ( 1 - scrollDistancePercentage );
				// end object settings
				
				parallaxScale = 0.8;
				parallaxOpacity = 0;
				parallaxOffset = -100;
				factor = 0;
				
				if( scrollTop > effectiveFrom )
				{
					factor = ( scrollTop - effectiveFrom ) / ( effectiveTo - effectiveFrom );
					factor = ( factor > 1 ? 1 : factor );
				}
				
				options = {
					opacity: 1,
					scale: 1,
					translateX: 0,
					translateY: 0
				};
				
				if( $( obj ).hasClass( "-opacity" ) )
				{
					options.opacity = 0 + factor;
				}
				
				if( $( obj ).hasClass( "-pop" ) )
				{
					options.scale = 0.8 + factor * 0.2;
				}
				
				if( $( obj ).hasClass( "-slide-left" ) )
				{
					options.translateX = ( -100 + factor * 100 ) * -1;
				}
				
				if( $( obj ).hasClass( "-slide-right" ) )
				{
					options.translateX = ( -100 + factor * 100 );
				}
				
				if( $( obj ).hasClass( "-slide-top" ) )
				{
					options.translateY = ( -100 + factor * 100 ) * -1;
				}
				
				if( $( obj ).hasClass( "-slide-bottom" ) )
				{
					options.translateY = ( -100 + factor * 100 );
				}
				
				$( obj ).css({
					webkitFilter: "opacity(" + options.opacity + ")",
					mozFilter: "opacity(" + options.opacity + ")",
					oFilter: "opacity(" + options.opacity + ")",
					msFilter: "opacity(" + options.opacity + ")",
					filter: "opacity(" + options.opacity + ")",
					
					webkitTransform: "translate3d( " + parseInt( options.translateX ) + "px, " + parseInt( options.translateY ) + "px, 0px ) scale(" + options.scale + ")",
					mozTransform: "translate3d( " + parseInt( options.translateX ) + "px, " + parseInt( options.translateY ) + "px, 0px ) scale(" + options.scale + ")",
					oTransform: "translate3d( " + parseInt( options.translateX ) + "px, " + parseInt( options.translateY ) + "px, 0px ) scale(" + options.scale + ")",
					msTransform: "translate3d( " + parseInt( options.translateX ) + "px, " + parseInt( options.translateY ) + "px, 0px ) scale(" + options.scale + ")",
					transform: "translate3d( " + parseInt( options.translateX ) + "px, " + parseInt( options.translateY ) + "px, 0px ) scale(" + options.scale + ")",
					
					transition: "all " + effectDuration + "ms ease-out"
				});
			});
			
			return;
			
			// TODO: Timeout for mobile devices
			if( this.options.useMobileTimeouts && this.lastScrollTop != scrollTop )
			{
				this.lastScrollTop = scrollTop;
				
				$( "body" ).stop();
				$( "body" ).animate({ float: "none" }, this.options.mobileTimeout, function()
				{
					$this.update();
				});
			}
		}
	});
	